custreamz_notebooks/streaming_regex_aggregations_haproxy.ipynb

```diff
@@ -0,0 +1,292 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**This is a cuStreamz job which streams HAProxy logs from Kafka, parses them, and performs some basic aggregations.**\n",
+    "\n",
+    "For this example, we will be demonstrating how to stream from Kafka. But one can also perform the same streaming operations reading from a text file which is being continuously written into."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "*An example HAProxy log would be of the following form:*\n",
+    "\n",
+    "'[haproxy@10.0.0.1] <134>May 29 19:08:36 haproxy[113498]: 45.26.605.15:38738 [29/May/2019:19:08:36.691] HTTPS:443~ HTTP_ProvisionManagers/mp3 4/5/0/1/1 200 6182 - - --NI 3/3/0/0/0 0/0 {|} \"GET /v2/serverinfo HTTP/1.1'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Streamz and cudf imports\n",
+    "from streamz import Stream\n",
+    "import cudf\n",
+    "from streamz.dataframe import DataFrame\n",
+    "import time"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Below is a helper function which implements RegEx parsing on the HAProxy logs, and then calculates the average backend response time for each batch.\n",
+    "\n",
+    "It also has timestamps to determine the time taken by each important phase of the stream processing — parsing and aggregations. These times are returned along with the average backend response time taken for each batch."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def haproxy_regex_aggregations(messages):\n",
+    "    \n",
+    "    preprocess_start_time = int(round(time.time()))\n",
+    "    \n",
+    "    size = len(messages)*len(messages[0]) \n",
+    "    num_rows = len(messages)\n",
+    "    json_input_string = \"\\n\".join([msg.decode('utf-8') for msg in messages])\n",
```

Comment:
If the messages are  `byte`  like objects, may be we can do something like : 
```python3
from io import BytesIO
import cudf

byte_string = b'sample bytes\nsample bytes'
df = cudf.read_csv(BytesIO(byte_string),header=None,names=['text'])
```

Reply:
1. My bad, each message in the Kafka topic is a JSON-encoded string (that's how most of Kratos streaming workloads are; hence, I wanted to conform to that) of the form: {'logline':<log>}. Will add this to the notebook, thanks a lot for pointing this out! 

2. Yes, since they're JSON-encoded strings.

3. I thought of trying this, but I am sure this won't work for this format of data.